hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/security/CompositeGroupsMapping.java

```diff
@@ -78,15 +79,13 @@
             user, provider.getClass().getSimpleName(), e.toString());
         LOG.debug("Stacktrace: ", e);
       }        
-      if (groups != null && ! groups.isEmpty()) {
+      if (!groups.isEmpty()) {
         groupSet.addAll(groups);
         if (!combined) break;
       }
     }
 
-    List<String> results = new ArrayList<String>(groupSet.size());
-    results.addAll(groupSet);
-    return results;
+    return Collections.unmodifiableList(new ArrayList<>(groupSet));
```

Comment:
Now, somebody calling getGroups() would start failing if it was adding new things to the list while that was not the case before. I've tried using inmodifiable before but is kind of dangerous.